cvprac/cvp_api.py

```diff
@@ -3665,39 +3665,130 @@ def change_control_stop(self, cc_id, notes=""):
+                cc_id (string): The ID for the new change control.
+                schedule_time (string): rfc3339 time format, e.g: 2021-12-23T02:07:00.0Z
+                notes (string): An optional note.
+            Returns:
+                response (dict): A dict that contains...
+                Ex: {"value":{"key":{"id":"5821c7c1-e276-4387-b60a"},
+                              "schedule":{"value":"2021-12-23T02:07:00Z",
+                                          "notes":"CC schedule via curl"}},
+                     "time":"2021-12-23T02:06:18.739965204Z"}
+        '''
+        payload = {
+                "key": {
+                    "id": cc_id
+                },
+                "schedule": {
+                    "value": schedule_time,
+                    "notes": notes
+                }
+        }
+        cc_url = '/api/resources/changecontrol/v1/ChangeControlConfig'
+        # For on-prem check the version as it is only supported from 2022.1.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion < 8.0:
+                self.log.warning(
+                     'Change Control Scheduling via Resource APIs are supported from 2022.1.0 or newer.')
+                return None
+        self.log.debug('v8 ' + str(cc_url) + ' ' + str(payload))
+        return self.clnt.post(cc_url, data=payload, timeout=self.request_timeout)
+
+    def device_decommissioning(self, device_id, request_id):
+        ''' Decommission a device using Resource APIs.
+            Supported versions: CVP 2021.3.0 or newer and CVaaS.
+            Args:
+                device_id (string): Serial Number of the device.
+                request_id (string): Key identifies the request to decommission the device.
+                    Recommended to generate uuid with str(uuid.uuid4()).
+            Returns:
+                response (dict): A dict that contains...
+                Ex: {'value': {'key': {'requestId': '4a4ba5a2-9886-4cd5-84d6-bdaf85a9f091'},
+                     'deviceId': 'BAD032986065E8DC14CBB6472EC314A6'},
+                     'time': '2022-02-12T02:58:30.765459650Z'}
+        '''
+        payload = {
+            "key": {
+                "request_id": request_id
+            },
+            "device_id": device_id
+        }
+        url = '/api/resources/inventory/v1/DeviceDecommissioningConfig'
+        # For on-prem check the version as it is only supported from 2021.3.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion < 7.0:
+                self.log.warning(
+                    'Decommissioning via Resource APIs are supported from 2021.3.0 or newer.')
+                return None
+        self.log.debug('v7 ' + str(url) + ' ' + str(payload))
+        return self.clnt.post(url, data=payload, timeout=self.request_timeout)
```

Comment:
Would be good to check if the device already exists before decommissioning.

Reply:
Another suggestion, to check if there are no other tasks/change_control currently running on the device because it leads to failure al though api is successful.

Reply:
I've added a check now using get_device_by_serial and print a warning log and return None if the result is an empty dict, is that the best way to handle this?

regarding checking if a device has a task assigned or not, there's no such API, and it's not scalable to go through each task and check if it's pending or not, it's much faster and easier to just try to decommission the device and then poll the state of the decommissioning via `device_decommissioning_status_get_one` and if it returns `Pending/In-progress tasks exist for the device` the user has to cancel the tasks

Reply:
Do we know what happens when you run the decommission API against a device that doesn't exist? Does it return an error or simply return nothing?